crates/stages/src/stages/execution.rs

```diff
@@ -190,8 +196,70 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
 
         let is_final_range = stage_progress == max_block;
         info!(target: "sync::stages::execution", stage_progress, is_final_range, "Stage iteration finished");
-        Ok(ExecOutput { checkpoint: StageCheckpoint::new(stage_progress), done: is_final_range })
+        Ok(ExecOutput {
+            checkpoint: StageCheckpoint::new(stage_progress)
+                .with_execution_stage_checkpoint(stage_checkpoint),
+            done: is_final_range,
+        })
+    }
+}
+
+fn execution_checkpoint<DB: Database>(
+    tx: &Transaction<'_, DB>,
+    start_block: BlockNumber,
+    max_block: BlockNumber,
+    checkpoint: StageCheckpoint,
+) -> Result<ExecutionCheckpoint, DatabaseError> {
+    Ok(match checkpoint.execution_stage_checkpoint() {
+        // If checkpoint block range fully matches our range,
+        // we take the previously used stage checkpoint as-is.
+        Some(stage_checkpoint @ ExecutionCheckpoint { block_range, .. })
+            if block_range == CheckpointBlockRange::from(start_block..=max_block) =>
+        {
+            stage_checkpoint
+        }
+        // If checkpoint block range precedes our range seamlessly, we take the previously used
+        // stage checkpoint and add the amount of gas from our range to the checkpoint total.
+        Some(ExecutionCheckpoint {
+            block_range: CheckpointBlockRange { to, .. },
+            progress: EntitiesCheckpoint { processed, total: Some(total) },
+        }) if to == start_block - 1 => ExecutionCheckpoint {
+            block_range: CheckpointBlockRange { from: start_block, to: max_block },
+            progress: EntitiesCheckpoint {
+                processed,
+                total: Some(total + calculate_gas_used_from_headers(tx, start_block..=max_block)?),
+            },
+        },
+        // Otherwise, we recalculate the whole stage checkpoint including the amount of gas
+        // already processed, if there's any.
+        _ => {
+            let processed = calculate_gas_used_from_headers(tx, 0..=start_block - 1)?;
+
+            ExecutionCheckpoint {
+                block_range: CheckpointBlockRange { from: start_block, to: max_block },
+                progress: EntitiesCheckpoint {
+                    processed,
+                    total: Some(
```

Comment:
might've missed it in previous PRs, is there a good reason for total to be an option

Reply:
hmm, it was needed for hashing stages when we were operating on different entity types in execution and unwind, but now it's not relevant anymore. Will remove in a separate PR, so this one won't get even bigger because of changes in other stages.

Reply:
sg

Reply:
https://github.com/paradigmxyz/reth/pull/2957 🫡